src/AppInstallerCommonCore/Reboot.cpp

```diff
@@ -0,0 +1,53 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+#include "pch.h"
+#include "Public/winget/Reboot.h"
+#include <Windows.h>
+
+namespace AppInstaller::Reboot
+{
+#ifndef AICLI_DISABLE_TEST_HOOKS
+    static bool* s_HasRebootPrivilegeResult_TestHook_Override = nullptr;
+
+    void TestHook_SetHasRebootPrivilegeResult_Override(bool* status)
+    {
+        s_HasRebootPrivilegeResult_TestHook_Override = status;
+    }
+#endif
+
+    bool HasRebootPrivilege()
```

Comment:
This function is not just checking for the presence of the privilege, but permanently adjusting the process token to enable it.

If there is need of it, we should have a function that *only* checks for the privilege in the token, not adjusting in.

Then, `IntiateReboot` should be the one responsible for enabling the privilege.  It can return value with more than two states to indicate no privilege, a failure to reboot, or success.  Additionally, the proper method for adjusting privileges in a multi-threaded process is to duplicate the token, adjust the duplicate's privilege, use it as the thread token, take privileged action, then remove the thread token.

Reply:
I made a mistake in the implementation. Turns out we don't need to check for the presence of the privilege. The token needs to be adjusted with the enabled shutdown privilege in order to enable shutdown. I have removed the privilege check and now just adjust the token privilege prior to initiating a reboot with the appropriate error logging if that fails.